changeset 7064:ecb5cd305b06

Core: disabled SO_REUSEPORT when testing config (ticket #1300).

When closing a socket with SO_REUSEPORT, Linux drops all connections
waiting in this socket's listen queue. Previously, it was believed to
only result in connection resets when reconfiguring nginx to use smaller
number of worker processes. It also results in connection resets during
configuration testing though.

Workaround is to avoid using SO_REUSEPORT when testing configuration.
It should prevent listening sockets from being created if a conflicting
socket already exists, while still preserving detection of other possible
errors. It should also cover UDP sockets.

The only downside of this approach seems to be that a configuration
testing won't be able to properly report the case when nginx was compiled
with SO_REUSEPORT, but the kernel is not able to set it. Such errors
will be reported on a real start instead.

--- a/src/core/ngx_connection.c
+++ b/src/core/ngx_connection.c
@@ -473,7 +473,7 @@
 
 #if (NGX_HAVE_REUSEPORT)
 
-            if (ls[i].reuseport) {
+            if (ls[i].reuseport && !ngx_test_config) {
                 int  reuseport;
 
                 reuseport = 1;
